### PR TITLE
Add upload study artifact api

### PR DESCRIPTION
**Pull request by gen740**
## Contributor License Agreement

This repository (``optuna-dashboard``) and [Goptuna](https://github.com/c-bata/goptuna) share common code.
This pull request may therefore be ported to Goptuna.
Make sure that you understand the consequences concerning licenses and check the box below if you accept the term before creating this pull request.

* [x] I agree this patch may be ported to [Goptuna](https://github.com/c-bata/goptuna) by other Goptuna contributors.

## Reference Issues/PRs

None.

## What does this implement/fix? Explain your changes.

- Add the `upload_study_artifact` api.
- Add the study artifact cards in the `StudyHistory` page.

スクリーンショット 2023-10-27 17 36 02
スクリーンショット 2023-10-27 17 36 07

### CODE REVIEW

**Comment by HideakiImamura** [comment]
Let me read this PR for my study. I will not merge the PR with my approval.

File: optuna_dashboard/artifact/_backend.py
Comment (HideakiImamura) [inline optuna_dashboard/artifact/_backend.py:149]: Is it possible to use `optuna.upload_artifact` in this function?

File: optuna_dashboard/artifact/_backend.py
Comment (HideakiImamura) [inline optuna_dashboard/artifact/_backend.py:201]: Is it possible to use `artifact_store.remove` in this function?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the PR. I have some early comments. PTAL.

File: optuna_dashboard/artifact/_backend.py
Comment (gen740) [reply optuna_dashboard/artifact/_backend.py:149]: Yes, it could. However, let's maintain the current changes for the moment. As @c-bata suggested, `upload_trial_artifact_api` could also be refactored to use `optuna.upload_artifact`. We should address these refinements in a follow-up PR.

File: optuna_dashboard/artifact/_backend.py
Comment (gen740) [reply optuna_dashboard/artifact/_backend.py:201]: Actually, it has already been used in this function.

File: optuna_dashboard/artifact/_backend.py
Comment (gen740) [reply optuna_dashboard/artifact/_backend.py:149]: FYI, the `optuna.upload_artifact` function accepts `Study`, `Trial`, or `FrozenTrial` objects, but it only utilizes `study_id` and `trial_id`. In the current implementation, we retrieve the study object from the storage object and then extract the `study_id` from it. This approach is somewhat redundant and inefficient.

File: optuna_dashboard/artifact/_backend.py
Comment (keisuke-umezawa) [inline optuna_dashboard/artifact/_backend.py:207]: [question] As study artifact is introduced in this PR, the metadata is always stored in `ARTIFACTS_ATTR_PREFIX + artifact_id`?

File: optuna_dashboard/artifact/_backend.py
Comment (keisuke-umezawa) [reply optuna_dashboard/artifact/_backend.py:149]: > We should address these refinements in a follow-up PR.

Can you create a follow-up PR or issue not to forget this discussion? It is ok after this PR is merged.

**Review by keisuke-umezawa — COMMENTED** [review]
Sorry for being late. I just checked python API part, and left comments. Regarding python API parts, it looks almost good to me other than my comments.
If you divide this PR into python API change and adding new UI, I can merge the PR for python API change immediately.

**Comment by keisuke-umezawa** [comment]
It is not relevant to this PR, but should we also need to delete meta data in `delete_all_artifacts`?

**Comment by keisuke-umezawa** [comment]
Also, it is not relevant to this PR or optunda-dashboard, but should we prepare API for deleting data and meta together like `delete_artifact` similar to `upload_artifact` in optuna itself?

File: optuna_dashboard/artifact/_backend.py
Comment (HideakiImamura) [reply optuna_dashboard/artifact/_backend.py:201]: Ops. Sorry for the confusion.

File: optuna_dashboard/ts/components/StudyHistory.tsx
Comment (HideakiImamura) [inline optuna_dashboard/ts/components/StudyHistory.tsx:175]: This is more of a question rather than a comment, but what are the differences between Grid and Grid2? I am having trouble finding documentation on Grid2.

File: optuna_dashboard/ts/components/StudyArtifactCards.tsx
Comment (HideakiImamura) [reply optuna_dashboard/ts/components/StudyArtifactCards.tsx:39]: I see, these values are designed to be the same as in TrialArtifactCards.tsx. If that's the case, it's understandable.

On another note, could we extract and reuse the common code from TrialArtifactCards.tsx?

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the update. I added some additional comments. PTAL

File: optuna_dashboard/artifact/_backend.py
Comment (gen740) [reply optuna_dashboard/artifact/_backend.py:207]: Yes, this behavior is the same as [optuna/artifacts/_upload.py](https://github.com/optuna/optuna/blob/f966055e159bd16093b679e7c3c316c9d2c0d2b5/optuna/artifacts/_upload.py#L82)

File: optuna_dashboard/ts/components/StudyHistory.tsx
Comment (gen740) [reply optuna_dashboard/ts/components/StudyHistory.tsx:175]: In the [document](https://mui.com/material-ui/react-grid2/#whats-changed),
```
import Grid from '@mui/material/Unstable_Grid2';
```
and
```
import Grid2 from '@mui/material/Unstable_Grid2';
```
appear to be the same. Currently, `optuna-dashboard` uses both `Grid` and `Unstable_Grid2`. Let's create an issue to unify them.

File: optuna_dashboard/ts/components/StudyArtifactCards.tsx
Comment (gen740) [reply optuna_dashboard/ts/components/StudyArtifactCards.tsx:31]: Yes, it's true. However, let me create the issue for refactoring these changes, including these changes.

File: optuna_dashboard/ts/components/StudyArtifactCards.tsx
Comment (gen740) [reply optuna_dashboard/ts/components/StudyArtifactCards.tsx:39]: I created the [follow-up PR](#692)

File: optuna_dashboard/artifact/_backend.py
Comment (gen740) [reply optuna_dashboard/artifact/_backend.py:149]: I created the issue #693.

File: optuna_dashboard/artifact/_backend.py
Comment (keisuke-umezawa) [reply optuna_dashboard/artifact/_backend.py:207]: In that case, can we remove those lines?

```
storage.set_study_system_attr(
            study_id, _dashboard_artifact_prefix(study_id) + artifact_id, json.dumps(None)
        )
```

**Comment by keisuke-umezawa** [comment]
I checked ui side and as a code, looks almost good to me. However, in my environment, I can only see blank study artifact without uploader. What should I do to show uploader?
スクリーンショット 2023-11-12 16 24 52

**Comment by gen740** [comment]
> However, in my environment, I can only see a blank study artifact without an uploader.

Good catch, that's because you did not pass the `artifact_store` instance to `run_server` (e.g., `run_server(storage, artifact_store=artifact_store)`).
Anyway, this is a bug because it should not appear when `artifact_store` isn't passed. Thanks!

File: optuna_dashboard/artifact/_backend.py
Comment (gen740) [reply optuna_dashboard/artifact/_backend.py:207]: It sounds good. I think I can remove that line from `delete_study_artifact`. However, I would not remove it from `delete_trial_artifact` due to backward compatibility.

**Review by keisuke-umezawa — APPROVED** [review]
LGTM!
Thank you for working on it.

**Pull request merged**
